### PR TITLE
MXNet -> Apple CoreML converter (Review-only. Don't merge)

### PR DESCRIPTION
**Pull request by pracheer**
This tool helps convert MXNet models into [Apple CoreML](https://developer.apple.com/documentation/coreml) format which can then be run on Apple devices.

This PR is currently for review only since we are in the process of figuring out the licensing issues with Apple. We got the initial code from them and they have handed over the maintenance and development of this tool to MXNet community from now on.

This has been tested against MXNet 0.10.0 as well as against the current master branch.

Example of a sample command to convert squeeze-net model. 
`python mxnet_coreml_converter.py --model-prefix='squeezenet_v1.1' --epoch=0 --input-shape='{"data":"3,224,224"}' --output-file="squeezenet_v11.mlmodel"`

This tools is currently able to convert Inception/Resnet/VGG/Squeezenet models. Support for more vision models as well as nlp models will be added later.

Unit tests have been added which test 1/ each layer individually 2/ some of the standard models from mxnet zoo against random data as well as on a subset of imagenet data. In all the tests we compare the accuracy of MXNet vs that of CoreML using the converted model.

@mli @piiswrong @madjam @arank @jiajiechen @indhub

### CODE REVIEW

File: tools/core_ml/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Convert MXNet models into Apple CoreML format.
+
+This tool helps convert MXNet models into [Apple CoreML](https://developer.apple.com/documentation/coreml) format which can then be run on Apple devices.
+
+## Installation
+In order to use this tool you need to have these installed:
+* mxnet 0.10.0. [Installation instructions](http://mxnet.io/get_started/install.html).
```
Comment (madjam) [inline tools/core_ml/README.md]: should this be version >= 0.10.0

File: tools/core_ml/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+* python 2.7
+* coremltools 0.4.0 (pip install coremltools)
+* yaml (pip install pyyaml)
+
+## How to use
+Let's say you want to use your MXNet model in an iPhone App. For the purpose of this example, let's say you want to use squeezenet-v1.1.
+
+1. Download the model into the directory where this converter resides. Squeezenet can be downloaded from [here](http://data.mxnet.io/models/imagenet/squeezenet/).
```
Comment (madjam) [inline tools/core_ml/README.md]: why copy into the directory where convertor resides? Why not have the convertor take as input model directory?

File: tools/core_ml/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+You could ask CoreML to pre-process the images before passing them through the model.
+
+```bash
+python mxnet_coreml_converter.py --model-prefix='squeezenet_v1.1' --epoch=0 --input-shape='{"data":"3,224,224"}' --pre-processing-arguments='{"red_bias":127,"blue_bias":117,"green_bias":103}' --output-file="squeezenet_v11.mlmodel"
+```
+
+
+## Currently supported
+### Models
+This is a (growing) list of standard MXNet models that can be successfully converted using the converter. This means that any other model that uses the similar operators as these models can also be successfully converted.
```
Comment (madjam) [inline tools/core_ml/README.md]: "...that uses the similar operators..." -> remove "the"

File: tools/core_ml/README.md (hunk 2 above)
Comment (zhreshold) [inline tools/core_ml/README.md]: I think mx.model.load_checkpoint can take whatever the path is, absolute/relative paths are ok.

Comment (madjam) [inline tools/core_ml/README.md]: Having to copy model to the convertor location is not a good idea. In some situations you may not even have permissions to do that. Can you look into the suggestion @zhreshold made above?

**Comment by szha** [comment]
@pracheer could you take a look and see if it can support mxnet.gluon.model_zoo.vision models.

**Comment by pracheer** [comment]
Superseded by https://github.com/apache/incubator-mxnet/pull/7438

**Comment by pracheer** [comment]
Previous pull request: https://github.com/apache/incubator-mxnet/pull/7337

**Pull request closed without merging**